sbnanalysis/ana/SBNOsc/NumuSelection.cxx

```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+    contained_in_FV = true;
+    l_contained_length = 0;
+    l_length = 0;
+    smeared_eccqe = 0;
+
+    auto const& lepton_track = mctrack_list.at(lepton_ind);
+    
+    // Get the length and determine if any point leaves the fiducial volume
+    TLorentzVector pos = lepton_track.Start().Position();
+    for (int i = 1; i < lepton_track.size(); i++) {
+      // update if track is contained
+      if (contained_in_FV) contained_in_FV = containedInFV(pos.Vect());
+      
+      // update length
+      l_contained_length += containedLength(lepton_track[i].Position().Vect(), pos.Vect(), _config.fiducial_volumes);
+      l_length += (lepton_track[i].Position().Vect() - pos.Vect()).Mag();
+      
+      pos = lepton_track[i].Position();
     }
+    
+    // determine if stopped in TPC
+    // sim::MCStep end = lepton_track.End();
+    // geoalgo::Point_t l_end_pos(end.Position().Vect());
+    // bool stop_in_tpc = _config.active_volume.Contain(l_end_pos);
+
+    // While we're here, get the smeared ECCQE
+    smeared_eccqe = ECCQE(lepton_track.Start().Momentum().Vect(), lepton_track.Start().E()/1000., 
+      _smear->Smear(lepton_track.Start().E()/1000., 14, contained_in_FV));
   }
+    
+  // get visible energy
+  // "Reconstruct" visible energy from tracks + showers
+  // Also do smearing
+  //
+  // Track energy and masses are in MeV
+  double visible_E = 0.;
+  double smeared_visible_E = 0.;
+  // first the tracks
+  for (auto const &mct: mctrack_list) {
+    if (isFromNuVertex(mctruth, mct)) {
+      // don't subtract mass for muon
+      double mass = (mct.PdgCode() == 13) ? 0:PDGMass(mct.PdgCode());
+      double this_visible_energy = mct.Start().E() - mass;
+      double this_smeared_visible_energy = this_visible_energy + _smear->Smear(this_visible_energy, mct.PdgCode(), contained_in_FV);
+      visible_E += this_visible_energy;
+      smeared_visible_E += this_smeared_visible_energy; 
+    }
+  }
+  // now the showers
+  for (auto const &mcs: mcshower_list) {
+    if (isFromNuVertex(mctruth, mcs)) {
+      double mass = PDGMass(mcs.PdgCode());
+      double this_visible_energy = mcs.Start().E() - mass;
+      double this_smeared_visible_energy = this_visible_energy + _smear->Smear(this_visible_energy, mcs.PdgCode(), contained_in_FV);
+      visible_E += this_visible_energy;
+      smeared_visible_E += this_smeared_visible_energy; 
+    }
+  }
+
+
+  // convert visible energies to GeV
+  return {contained_in_FV, l_contained_length, l_length, visible_E / 1000., smeared_visible_E / 1000., smeared_eccqe}; 
+}
+
+std::vector<bool> NumuSelection::Select(const gallery::Event& ev, const simb::MCTruth& mctruth, unsigned truth_ind, const NumuSelection::NuMuInteraction &intInfo) {
+  // get the neutrino
+  const simb::MCNeutrino& nu = mctruth.GetNeutrino();
 
-  bool selected = !reco.empty();
+  // select true CC events
+  bool pass_true_CC = (nu.CCNC() == simb::kCC && (nu.Mode() == 0 || nu.Mode() == 10) && nu.Nu().PdgCode() == 14);
```

Comment:
It looks like this eventually is a requirement for the event to pass the selection, which I think is too strict -- the selected sample should include non-CCQE backgrounds like pion mis-IDs or CCQE-like topologies generated through FSI (pion absorption, say), which this will cut.

Reply:
Removed this cut and added in pion track confusion.